File: IBS/ModelRegenerate.py
import time


class ModelRegenerate:
    def __init__(self, Temp_path: str, model_path: str, model_name: str, t: int, *args):
        self.Temp_path = Temp_path
        self.model_path = model_path
        self.model_name = model_name
        self.t = t
        self.parameterSet = args

    def command_is_ready(self):
        with open(self.Temp_path + "CreoCommandReady.txt", "w") as file:
            file.write("")
        file.close()
        time.sleep(self.t)

    def model_load(self):
        with open(self.Temp_path + "CreoCommand.txt", "w") as file:
            file.write("LOAD_MODEL" + "\n" + self.model_path + self.model_name)
        file.close()
        time.sleep(self.t)
        self.command_is_ready()

    def SET_PARAMT_REGEN(self):
        with open(self.Temp_path + "CreoCommand.txt", "w") as file:
            file.write("SET_PARAMT_REGEN" + "\n")
        file.close()
        time.sleep(self.t)
        with open(self.Temp_path + "CreoCommand.txt", "a") as file:
            num = int(len(self.parameterSet) / 2)
            for i in range(num):
                file.write(str(self.parameterSet[i]) + "=" + str(self.parameterSet[i + num]) + '\n')
        file.close()
        time.sleep(self.t)
        self.command_is_ready()
    # def model_parameter_regenerate(self, para_name, para_value):

    def model_save(self):
        with open(self.Temp_path + "CreoCommand.txt", "w") as file:
            file.write("SAVE_MODEL" + "\n")
        file.close()
        time.sleep(self.t)
        self.command_is_ready()

    def model_export_sat(self):
        with open(self.Temp_path + "CreoCommand.txt", "w") as file:
            file.write("EXPORT_MODEL_SAT" + "\n" + self.model_path + self.model_name)
        file.close()
        time.sleep(self.t)
        self.command_is_ready()

    def stop(self):
        with open(Temp_path + "CreoCommand.txt", "w") as file:
            file.write("STOP" + "\n")
        file.close()
        time.sleep(self.t)
        self.command_is_ready()

    def execute(self):
        self.model_load()
        self.SET_PARAMT_REGEN()
        self.model_save()
        self.model_export_sat()

File: ModelRegenerate/ModelRegenerate.m
classdef ModelRegenerate
    properties
        creobridgePath
        tempFolderPath
        modelPath
        pauseTime
        parameters
        value
        modelName
        parameterAndValue
    end
    
    methods
        function obj = ModelRegenerate(creobridgePath, tempFolderPath, ...
                modelPath, modelName, pauseTime, parameters, value)
            obj.creobridgePath = creobridgePath;
            obj.tempFolderPath = tempFolderPath;
            obj.modelPath = modelPath;
            obj.modelName = modelName;
            obj.pauseTime = pauseTime;
            obj.parameters = parameters;
            obj.value = value;
            for c = 1:length(obj.parameters)
                obj.parameterAndValue = sprintf('%s%s=%f;', obj.parameterAndValue, obj.parameters(c), obj.value(c));
            end
        end
        
        function creoCommandReady(obj)
            fid = fopen([obj.tempFolderPath, '\CreoCommandReady.txt'], 'w');
            fprintf(fid, '%s', '');
            fclose(fid);
            pause(obj.pauseTime);
        end
        
        function loadModel(obj)
            fid = fopen([obj.tempFolderPath, '\CreoCommand.txt'], 'w');
            fprintf(fid, '%s\n%s\\%s%s', 'LOAD_MODEL', obj.modelPath, obj.modelName, '.prt');
            fclose(fid);
            pause(obj.pauseTime);
            creoCommandReady(obj);
        end
        
        function setParameters(obj)
            fid = fopen([obj.tempFolderPath, '\CreoCommand.txt'], 'w');
            fprintf(fid, '%s\n', 'SET_PARAMT_REGEN');
            for c = 1:length(obj.parameters)
                fprintf(fid, '%s=%f\n', obj.parameters(c), obj.value(c));
                %obj.parameterAndValue = [obj.parameterAndValue obj.parameters(c) obj.value(c)];
                obj.parameterAndValue = [1];
            end
            fclose(fid);
            pause(obj.pauseTime);
            creoCommandReady(obj);
        end
        
        function saveModel(obj)
            fid = fopen([obj.tempFolderPath, '\CreoCommand.txt'], 'w');
            fprintf(fid, '%s', 'SAVE_MODEL');
            fclose(fid);
            pause(obj.pauseTime);
            creoCommandReady(obj);
        end
        
        function saveExtraModel(obj)
            path = sprintf('%s%s%s', obj.modelPath, '\', obj.parameterAndValue);
            mkdir(path);
            fid = fopen([obj.tempFolderPath, '\CreoCommand.txt'], 'w');
            fprintf(fid, '%s\n%s', 'EXPORT_MODEL_CREO', path);
            fclose(fid);
            pause(obj.pauseTime);
            creoCommandReady(obj);
        end
        
        function exportModel(obj)
            fid = fopen([obj.tempFolderPath, '\CreoCommand.txt'], 'w');
            fprintf(fid, '%s\n%s\\%s%s', 'EXPORT_MODEL_SAT', obj.modelPath, obj.modelName, '.sat');
            fclose(fid);
            pause(obj.pauseTime);
            creoCommandReady(obj);
        end
        
        function stop(obj)
            fid = fopen([obj.tempFolderPath, '\CreoCommand.txt'], 'w');
            fprintf(fid, '%s', 'STOP');
            fclose(fid);
            pause(obj.pauseTime);
            creoCommandReady(obj);
        end
        
        function execute(obj)
            obj.loadModel();
            obj.setParameters();
            obj.saveModel();
            obj.saveExtraModel();
            obj.exportModel();
        end
    end
end

